### PR TITLE
Advanced Entity Queries

### PR DESCRIPTION
**Pull request by dmdashenkov**
This PR brings the support of the advanced Entity Queries to __gae-java__.

Also, this PR includes renaming the base package of the library from `org.spine3` to `io.spine`.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: gcd/src/main/java/org/spine3/server/storage/datastore/ColumnHandler.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import org.spine3.protobuf.TypeConverter;
+import org.spine3.server.entity.storage.Column;
+import org.spine3.server.entity.storage.ColumnTypeRegistry;
+import org.spine3.server.storage.datastore.type.DatastoreColumnType;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * A wrapper above the {@link ColumnTypeRegistry} preforming the transformations defined by
```
Comment (armiol) [inline gcd/src/main/java/org/spine3/server/storage/datastore/ColumnHandler.java]: `preforming` is a typo.

And it's a transformer, not a wrapper (which BTW is just some implementation details, that we can omit at all).

File: gcd/src/main/java/org/spine3/server/storage/datastore/ColumnHandler.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ */
+final class ColumnHandler {
+
+    private final ColumnTypeRegistry<? extends DatastoreColumnType> registry;
+
+    /**
+     * Creates a new instance of {@code ColumnHandler} on top of the given
+     * {@link ColumnTypeRegistry}.
+     */
+    static ColumnHandler wrap(ColumnTypeRegistry<? extends DatastoreColumnType> registry) {
```
Comment (armiol) [inline gcd/src/main/java/org/spine3/server/storage/datastore/ColumnHandler.java]: I'd call it `of`.

File: gcd/src/main/java/org/spine3/server/storage/datastore/ColumnHandler.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        checkNotNull(column);
+        checkNotNull(columnFilter);
+
+        final DatastoreColumnType type = registry.get(column);
+        checkArgument(type != null, "Column of unknown type: %s.", column);
+
+        final Any filterValue = columnFilter.getValue();
+        final Object filterValueUnpacked = TypeConverter.toObject(filterValue, column.getType());
+
+        @SuppressWarnings("unchecked") // Concrete type is unknown on compile time.
```
Comment (armiol) [inline gcd/src/main/java/org/spine3/server/storage/datastore/ColumnHandler.java]: I would say it's possible to introduce a generic parameter just for this method. It's still going to be "unchecked", but at least we'll know that the cast is performed to the exact type requested.

Let's discuss.

File: gcd/src/main/java/org/spine3/server/storage/datastore/DsFilters.java [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+    /**
+     * Matches the {@link CompositeQueryParameter} instances joined by
+     * the {@linkplain org.spine3.client.CompositeColumnFilter.CompositeOperator#ALL conjunctive}
+     * operator.
+     */
+    private static final Predicate<CompositeQueryParameter> isConjunctive =
+            new Predicate<CompositeQueryParameter>() {
+                @Override
+                public boolean apply(@Nullable CompositeQueryParameter input) {
+                    return input != null && input.getOperator() == ALL;
```
Comment (armiol) [inline gcd/src/main/java/org/spine3/server/storage/datastore/DsFilters.java]: To me, the `input` being `null` is an `IllegalArgumentException`.

File: gcd/src/main/java/org/spine3/server/storage/datastore/DsFilters.java [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+     */
+    private static final Predicate<CompositeQueryParameter> isDisjunctive = not(isConjunctive);
+
+    private DsFilters() {
+        // Prevent utility class fromm being initialized.
+    }
+
+    /**
+     * Converts the given {@link CompositeQueryParameter} instances into the Datastore
+     * {@link Filter}s.
```
Comment (armiol) [inline gcd/src/main/java/org/spine3/server/storage/datastore/DsFilters.java]: `{@linkplain Filter Filters}`.

File: gcd/src/main/java/org/spine3/server/storage/datastore/DsFilters.java [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+     * <p>The returned {@code Collection} contains the same predicate as the passed
+     * {@code Collection} of query parameters. The difference is in
+     * <ul>
+     *     <li>type: from Spine {@link CompositeQueryParameter} to Datastore {@link Filter};
+     *     <li>format: all the resulting {@link Filter}s are conjunctive (as Datastore does not
+     *         support disjunction) and are related with each other in the disjunctive way.
+     * </ul>
+     *
+     * <p><i>Example:</i>
+     * <p>Given query predicates {@code p1}, {@code p2}, {@code p3}, {@code p4}, {@code p5} passed
```
Comment (armiol) [inline gcd/src/main/java/org/spine3/server/storage/datastore/DsFilters.java]: Please allow an empty line before this `<p>`.

Also this example paragraph should be formatted better. It's a bit difficult to read ATM.

File: gcd/src/main/java/org/spine3/server/storage/datastore/ColumnHandler.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * A wrapper above the {@link ColumnTypeRegistry} preforming the transformations defined by
+ * the {@link org.spine3.server.entity.storage.ColumnType ColumnType} interface.
+ *
+ * @author Dmytro Dashenkov
+ */
+final class ColumnHandler {
```
Comment (armiol) [inline gcd/src/main/java/org/spine3/server/storage/datastore/ColumnHandler.java]: We need a new name for this class. This isn't a handler by any means.

**Review by armiol — CHANGES_REQUESTED** [review]
@dmdashenkov LGTM except for some minor comments.

Also, please review the version number before actual merge.

Any comments that you may have let's discuss vocally.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

File: gcd/src/main/java/org/spine3/server/storage/datastore/ColumnTypeConverter.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * A wrapper above the {@link ColumnTypeRegistry} performing the transformations defined by
+ * the {@link org.spine3.server.entity.storage.ColumnType ColumnType} interface.
+ *
+ * @author Dmytro Dashenkov
+ */
+final class ColumnTypeConverter {
```
Comment (armiol) [inline gcd/src/main/java/org/spine3/server/storage/datastore/ColumnTypeConverter.java]: `ColumnValueAdapter`. Let's update the Javadocs accordingly.

File: gcd/src/main/java/org/spine3/server/storage/datastore/DsFilters.java [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+        public int hashCode() {
+            return 0;
+        }
+    }
+
+    /**
+     * A functional interface defining the operation of precessing a conjunction group found by
+     * the {@linkplain #fromParams parenthesis opening algorithm}.
+     */
+    private interface ConjunctionProcessor {
```
Comment (armiol) [inline gcd/src/main/java/org/spine3/server/storage/datastore/DsFilters.java]: In fact this is a `Function<Queue<ColumnFilterNode>, Void>`.

File: gcd/src/main/java/org/spine3/server/storage/datastore/DsFilters.java [hunk 10]
```diff
@@ -0,0 +1,472 @@
...
+     *
+     * <p>The processor polls all the {@link ColumnFilterNode} instances, transforms them into
+     * {@link Filter} instances and merges with
+     * the {@link com.google.cloud.datastore.StructuredQuery.CompositeFilter#and
+     * StructuredQuery.CompositeFilter.and()} operation. The result is then collected into
+     * {@code Collection} passed on the instance creation.
+     *
+     * <p>The {@link Queue} passed to the function becomes empty after the processing.
+     */
+    private static class ColumnFilterShrinker implements ConjunctionProcessor {
```
Comment (armiol) [inline gcd/src/main/java/org/spine3/server/storage/datastore/DsFilters.java]: I think it's `ColumnFilterReducer`, judging on the documentation.

**Review by armiol — CHANGES_REQUESTED** [review]
@dmdashenkov LGTM except for naming.

File: gcd/src/main/java/org/spine3/server/storage/datastore/DsFilters.java (hunk 9 above)
Comment (dmdashenkov) [reply gcd/src/main/java/org/spine3/server/storage/datastore/DsFilters.java]: This doesn't look a proper usage of `Function` to me.
Moreover, the custom type provides a description of the usage of the type instances and `Function<Queue<Node>, Void>` says nothing about both the function itself and the parameter.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: gcd/src/main/java/io/spine/server/storage/datastore/DsFilters.java [hunk 11]
```diff
@@ -0,0 +1,472 @@
...
+
+        @Override
+        public int hashCode() {
+            return 0;
+        }
+    }
+
+    /**
+     * A functional interface defining the operation of precessing a conjunction group found by
+     * the {@linkplain #fromParams parenthesis opening algorithm}.
```
Comment (armiol) [inline gcd/src/main/java/io/spine/server/storage/datastore/DsFilters.java]: Let's rephrase this. `parenthesis opening` isn't what it is called. It's called `multiply brackets`, `expand brackets`, `simplify brackets` etc.

File: gcd/src/main/java/io/spine/server/storage/datastore/DsFilters.java [hunk 12]
```diff
@@ -0,0 +1,472 @@
...
+     */
+    private interface ConjunctionProcessor {
+
+        /**
+         * Processes the found conjunction.
```
Comment (armiol) [inline gcd/src/main/java/io/spine/server/storage/datastore/DsFilters.java]: This isn't good enough either. The description doesn't really explain why there is a `Queue<...>`, not just the "found conjunction".

**Review by armiol — CHANGES_REQUESTED** [review]
@dmdashenkov please see my comments, let's discuss.

**Comment by dmdashenkov** [comment]
@armiol, PTAL at the docs change.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

**Review by armiol — APPROVED** [review]
@dmdashenkov LGTM.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: README.md [hunk 13]
```diff
@@ -6,7 +6,51 @@
...
+
+This section describes the main aspects of using the library.
+
+#### Datastore indexes
+
+To work properly, datastore requires to configure the indexes. For the guide, visit [Google Cloud Platform Docs](https://cloud.google.com/datastore/docs/tools/indexconfig).
+
+##### Spine internal indexes
+
+To use `gae-java`, you should configure the Datastore indexes for the Spine internal record types.
```
Comment (armiol) [inline README.md]:
```
#### Datastore indexes

In order to run the application built on top of `gae-java`, Datastore instance requires some preliminary configuration. In particular, the indexes for the Spine internal record types should be set. 

The configuration file is located at `./gcd/config/index.yaml`. 

Please see the [Google Cloud Platform documentation](https://cloud.google.com/datastore/docs/tools/indexconfig) for the details.

##### Custom indexes

It is possible to store some of the Spine `Entity` fields in separate Datastore kind fields. Such an approach is useful to optimize read-side querying. In this case more Datastore indexes may be created.

__Example:__

...
```

**Pull request merged**